File: largestSumOfAverages/largestSumOfAverages.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    double largestSumOfAverages(vector<int>& A, int K) {
        vector<vector<double> > dp(A.size(), vector<double>(K + 1, -1));
        return dfs(A, K, A.size() - 1, dp);
    }
    // 0 ~ index is a sub-problem
    double dfs(vector<int>& A, int K, int index, vector<vector<double> >& dp) {
        if (dp[index][K] != -1)
            return dp[index][K];
        if (K == 1) {
            double sum = 0.0;
            for (int i = 0; i <= index; i++)
                sum += A[i];
            return dp[index][K] = sum / (index + 1);
        }
        if (index + 1 < K)
            return dp[index][K] = 0.0;
        double res = 0.0, lastPart = A[index];
        for (int i = index - 1; i >= 0; i--) {
            res = max (res, (lastPart / (index - i) + dfs(A, K - 1, i, dp)));
            lastPart += A[i];
        }
        return dp[index][K] = res;
    }
};

int main() {
    Solution s;
    vector<int> A = {9,1,2,3,9};
    int K = 3;
    cout << s.largestSumOfAverages(A, K) << endl;
    return 0;
}

File: largestSumOfAverages/largestSumOfAverages.java
class Solution {
    public double largestSumOfAverages(int[] nums, int k) {
        double[][] dp = new double[nums.length + 1][k + 1];
        dp[0][0] = 0.0;
        double sum0I = 0.0;
        for (int i = 1; i <= nums.length; i++) {
            sum0I += nums[i - 1];
            dp[i][1] = sum0I / i;
        }
        for (int j = 2; j <= k; j++) {
            for (int i = 1; i <= nums.length; i++) {           
                double sumXI = 0.0;
                for (int x = i - 1; x >= 0; x--) {
                    sumXI += nums[x];
                    dp[i][j] = Math.max(dp[i][j], dp[x][j - 1] + sumXI / (i - x));
                }
            }
        }
        return dp[nums.length][k];
    }
}
